plan/gen_physical_plans.go

```diff
@@ -271,32 +280,116 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
-		newInnerJoinKeys := make([]*expression.Column, len(innerJoinKeys))
-		for i, offset := range matchedOffsets {
-			newOuterJoinKeys[i] = outerJoinKeys[offset]
-			newInnerJoinKeys[i] = innerJoinKeys[offset]
+		// We choose the index by the number of used columns of the range, the much the better.
+		// Notice that there may be the cases like `t1.a=t2.a and b > 2 and b < 1`. So ranges can be nil though the conditions are valid.
+		// And obviously when the range is nil, we don't need index join.
+		if len(ranges) > 0 || len(ranges[0].LowVal) > maxUsedCols {
+			bestIndexInfo = indexInfo
+			maxUsedCols = len(ranges[0].LowVal)
+			rangesOfBest = ranges
+			remainedOfBest = remained
+
+			idxInIndex = idxInIndex[:0]
+			for key, val := range offsetsMap {
+				if val == -1 {
+					continue
+				}
+				idxInIndex = append(idxInIndex, key)
+			}
+			for i, idx := range idxInIndex {
+				pos := offsetsMap[idx]
+				orderedOuterKeys[i] = outerJoinKeys[pos]
+				orderedInnerKeys[i] = innerJoinKeys[pos]
+			}
 		}
-		outerJoinKeys = newOuterJoinKeys
-		innerJoinKeys = newInnerJoinKeys
-		break
 	}
-	if usedIndexInfo != nil {
-		innerPlan := x.forceToIndexScan(usedIndexInfo)
-		return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan)
+	if bestIndexInfo != nil {
+		innerPlan := x.forceToIndexScan(bestIndexInfo, remainedOfBest)
+		return p.constructIndexJoin(prop, orderedInnerKeys, orderedOuterKeys, outerIdx, innerPlan, rangesOfBest, idxInIndex)
 	}
 	return nil
 }
 
+// checkIndexForIndexJoin checks whether this index can be used for building index join and return enough information to
+// determine whether this index is better and how to construct the index join.
+func (p *LogicalJoin) checkIndexForIndexJoin(indexInfo *model.IndexInfo, innerPlan *DataSource, innerJoinKeys []*expression.Column) (ok bool,
+	remained []expression.Expression, indexRanges []*ranger.IndexRange, offsetsMap []int) {
+	idxCols, colLengths := expression.IndexInfo2Cols(innerPlan.Schema().Columns, indexInfo)
+	if len(idxCols) == 0 {
+		return false, nil, nil, nil
+	}
+
+	offsetsMap = joinKeysMatchIndex(innerJoinKeys, indexInfo)
+	if offsetsMap == nil {
+		return false, nil, nil, nil
+	}
+	maxIndexColIdx := 0
+	for i := len(indexInfo.Columns) - 1; i >= 0; i-- {
+		if offsetsMap[i] != -1 {
+			maxIndexColIdx = i
+			break
+		}
+	}
+
+	// After predicate push down, the one side conditions of join must be the conditions that cannot be pushed down so cannot
+	// to calculate range either. So we only need the innerPlan.pushedDownConds and the eq conditions that we generate.
+	// TODO: There may be a selection that block the index join.
+	conds := make([]expression.Expression, 0, len(offsetsMap)+len(innerPlan.pushedDownConds))
+	// Construct a fake equal expression for calculate the range.
+	for _, key := range innerJoinKeys {
+		fakeConstant := expression.One.Clone().(*expression.Constant)
+		fakeConstant.RetType = types.NewFieldType(key.RetType.Tp)
```

Comment:
Is it safe to use the constant value one? It may not be same as the key column's type.

Reply:
After i check, one can't fit `MysqlDuration` and `MysqlEnum`. I'm checking `table.GetZeroValue()`

Reply:
Ok, `table.GetZeroValue` doesn't meet the needs. And one can convert to `MysqlEnum` safely(see `ParseEnumValue` in `types/enum.go`). And column's type won't be `MysqlDuration`. So it's ok to use `one`.